lib/puma/cluster.rb

```diff
@@ -483,17 +496,23 @@ def run
                     debug_loaded_extensions("Loaded Extensions - master:") if @log_writer.debug?
                     booted = true
                   end
+                when "i"
+                  if @idle_workers[pid]
+                    @idle_workers.delete pid
```

Comment:
Hey,

_FYI I am new to Puma and may not fully understand the purpose of each possible value of `req`._

I had a bit of trouble from a newby perspective to understand why receiving `i` would make you remove it from the `@idle_workers`. I know understand that it means the worker is active so you want to actually remove it from the `@idle_workers` which make total sense.

Nevertheless these idle-me and idle-me-out requests require to read another file to fully understand the logic behind it.

Maybe to ease the reading of it we could:
1. Use a dedicated `req` letter to remove the `pid` from the `@idle_workers`, such as `a` for `active`?
2. Would it make sense to hide the technical `req` letters behind an enum such as:
```rb
module RequestType
  IDLE = "i"
  BOOT = "b"
  ACTIVE = "a"
  TERM = "t"
  EXTERNAL_TERM = "e"
end
```

Reply:
Great question and suggestions.

This is simply the prefix for the PID written to the pipe [here](https://github.com/joshuay03/puma/blob/2bd0b49b2b9a58bd35ee73add1da038f61815100/lib/puma/server.rb#L352). I didn't find it very intuitive when working on this either, and my original implementation had a two letter prefix (one for enable, one for disable) which forced me to update the rest of these "actions" to match (`i0`, `i1`, `te` etc.). The PR ended up with a few too many changes which would have made it harder to review so I went with a single "toggle" since the idle state in `@idle_workers` will be synchronous on a per-worker basis.

I do agree though that we can clean this up a bit and I'm a fan of your suggestion, but imo those changes should be in a separate PR.

Reply:
Actioning in #3318